sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java

```diff
@@ -336,4 +336,62 @@ public UnsafeArrayData copy() {
     arrayCopy.pointTo(arrayDataCopy, Platform.BYTE_ARRAY_OFFSET, sizeInBytes);
     return arrayCopy;
   }
+
+  public int[] toPrimitiveIntArray() {
+    int[] result = new int[numElements];
+    Platform.copyMemory(baseObject, baseOffset + 4 + 4 * numElements,
+      result, Platform.INT_ARRAY_OFFSET, 4 * numElements);
+    return result;
+  }
+
+  public double[] toPrimitiveDoubleArray() {
+    double[] result = new double[numElements];
+    Platform.copyMemory(baseObject, baseOffset + 4 + 4 * numElements,
+      result, Platform.DOUBLE_ARRAY_OFFSET, 8 * numElements);
+    return result;
+  }
+
+  public static UnsafeArrayData fromPrimitiveArray(int[] arr) {
+    int offsetRegionSize = 4 * arr.length;
+    int valueRegionSize = 4 * arr.length;
+    int totalSize = 4 + offsetRegionSize + valueRegionSize;
+    byte[] data = new byte[totalSize];
+
+    Platform.putInt(data, Platform.BYTE_ARRAY_OFFSET, arr.length);
+
+    int elementOffsetStart = 4 + offsetRegionSize;
+    for (int i = 0; i < arr.length; i++) {
+      Platform.putInt(data, Platform.BYTE_ARRAY_OFFSET + 4 + i * 4, elementOffsetStart + i * 4);
```

Comment:
This has two multiplications per element, which is expensive. Try:

~~~java
int regionOffset = Platform.BYTE_ARRAY_OFFSET + 4
int elementOffset = elementOffsetStart
for (...) {
  Platform.putInt(data, regionOffset, elementOffset);
  regionOffset += 4;
  elementOffset += 4;
}
~~~